internal/locate/region_request.go

```diff
@@ -1063,6 +1204,135 @@ func (s *sendReqState) send(bo *retry.Backoffer, req *tikvrpc.Request, timeout t
+		} else {
+			req.AccessLocation = kv.AccessCrossZone
+		}
+	}
+	req.Context.ClusterId = s.vars.rpcCtx.ClusterID
+	if req.InputRequestSource != "" && s.replicaSelector != nil {
+		patchRequestSource(req, s.replicaSelector.replicaType())
+	}
+	if s.vars.err = tikvrpc.SetContextNoAttach(req, s.vars.rpcCtx.Meta, s.vars.rpcCtx.Peer); s.vars.err != nil {
+		return false
+	}
+
+	// Count the replica number as the RU cost factor.
+	req.ReplicaNumber = 1
+	if s.vars.rpcCtx.Meta != nil && len(s.vars.rpcCtx.Meta.GetPeers()) > 0 {
+		req.ReplicaNumber = 0
+		for _, peer := range s.vars.rpcCtx.Meta.GetPeers() {
+			role := peer.GetRole()
+			if role == metapb.PeerRole_Voter || role == metapb.PeerRole_Learner {
+				req.ReplicaNumber++
+			}
+		}
+	}
+
+	return true
+}
+
+// handleAsyncResponse handles the response of an async request.
+func (s *sendReqState) handleAsyncResponse(start time.Time, canceled bool, resp *tikvrpc.Response, err error, cancels ...context.CancelFunc) (done bool) {
+	if len(cancels) > 0 {
+		defer func() {
+			for i := len(cancels) - 1; i >= 0; i-- {
+				cancels[i]()
+			}
+		}()
+	}
+	s.vars.resp, s.vars.err = resp, err
+	req := s.args.req
+	rpcDuration := time.Since(start)
+	if s.replicaSelector != nil {
+		recordAttemptedTime(s.replicaSelector, rpcDuration)
+	}
+	if s.Stats != nil {
+		s.Stats.RecordRPCRuntimeStats(req.Type, rpcDuration)
+	}
+	if s.vars.rpcCtx.Store != nil && req.ReplicaReadType == kv.ReplicaReadPreferLeader && !util.IsInternalRequest(req.RequestSource) {
+		s.vars.rpcCtx.Store.healthStatus.recordClientSideSlowScoreStat(rpcDuration)
+	}
+	if s.vars.rpcCtx.ProxyStore != nil {
+		fromStore := strconv.FormatUint(s.vars.rpcCtx.ProxyStore.storeID, 10)
+		toStore := strconv.FormatUint(s.vars.rpcCtx.Store.storeID, 10)
+		result := "ok"
+		if s.vars.err != nil {
+			result = "fail"
+		}
+		metrics.TiKVForwardRequestCounter.WithLabelValues(fromStore, toStore, req.Type.String(), result).Inc()
+	}
+
+	if err := s.vars.err; err != nil {
+		if isRPCError(err) {
+			s.rpcError = err
+		}
+		if s.Stats != nil {
+			errStr := getErrMsg(err)
+			s.Stats.RecordRPCErrorStats(errStr)
+			s.recordRPCAccessInfo(req, s.vars.rpcCtx, errStr)
+		}
+		if canceled {
+			metrics.TiKVRPCErrorCounter.WithLabelValues("context-canceled", storeIDLabel(s.vars.rpcCtx)).Inc()
+			return true
+		}
+		return false
+	}
+
+	s.vars.regionErr, s.vars.err = s.vars.resp.GetRegionError()
+	if s.vars.err != nil {
+		s.vars.rpcCtx, s.vars.resp = nil, nil
+		return true
+	} else if s.vars.regionErr != nil {
+		// need to handle region error
+		return false
+	}
+
+	if s.replicaSelector != nil {
+		s.replicaSelector.onSendSuccess(req)
+	}
+
+	return true
+}
+
+// toResponesExt converts the state to a ResponseExt .
+func (s *sendReqState) toResponesExt() (*tikvrpc.ResponseExt, error) {
+	if s.vars.err != nil {
+		return nil, s.vars.err
+	}
+	if s.vars.resp == nil {
+		return nil, errors.New("invalid state: response is nil")
+	}
```

Comment:
Should we also assert `s.vars.regionErr == nil` here?

Reply:
No, `toResponseExt` might be called with `.vars.regionErr` when the region error cannot be resolved by retry.